File: articles/networks-systems-biological-complexity/r/food_web_summary.R
# Ecological food-web summary.

data_path <- file.path("data", "food_web_edges.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "food_web_edges.csv")
}

edges <- read.csv(data_path)

nodes <- sort(unique(c(edges$source, edges$target)))

degree_summary <- data.frame(
  node = nodes,
  degree = sapply(nodes, function(node) {
    sum(edges$source == node | edges$target == node)
  })
)

network_summary <- data.frame(
  n_nodes = length(nodes),
  n_edges = nrow(edges),
  density = (2 * nrow(edges)) / (length(nodes) * (length(nodes) - 1)),
  mean_degree = mean(degree_summary$degree),
  max_degree = max(degree_summary$degree)
)

module_counts <- aggregate(source ~ module, data = edges, FUN = length)
names(module_counts)[names(module_counts) == "source"] <- "n_edges"

print(round(network_summary, 5))
print(module_counts[order(-module_counts$n_edges), ])
print(degree_summary[order(-degree_summary$degree), ])

File: articles/networks-systems-biological-complexity/python/food_web_summary.py
"""
Ecological food-web scaffold.

Run:
    python python/food_web_summary.py
"""

from pathlib import Path

import pandas as pd

from network_complexity_core import adjacency_matrix, degree_table, network_summary


ARTICLE_DIR = Path(__file__).resolve().parents[1]
EDGE_PATH = ARTICLE_DIR / "data" / "food_web_edges.csv"


def main() -> None:
    edges = pd.read_csv(EDGE_PATH)
    nodes, adjacency = adjacency_matrix(edges)

    summary = network_summary(nodes, adjacency)
    degree_df = degree_table(nodes, adjacency)

    module_counts = (
        edges.groupby("module")
        .size()
        .reset_index(name="n_edges")
        .sort_values("n_edges", ascending=False)
    )

    print(pd.DataFrame([summary.__dict__]).round(5).to_string(index=False))
    print(module_counts.to_string(index=False))
    print(degree_df.sort_values("degree", ascending=False).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
